contracts/libs/0.8.x/MerkleLib.sol

```diff
@@ -0,0 +1,101 @@
+// SPDX-License-Identifier: LGPL-3.0-only
+// Created By: Art Blocks Inc.
+
+import "@openzeppelin-4.7/contracts/utils/cryptography/MerkleProof.sol";
+
+pragma solidity ^0.8.0;
+
+/**
+ * @title Art Blocks Merkle Library
+ * @notice [TO FILL OUT]
+ * @author Art Blocks Inc.
+ */
+
+library MerkleLib {
+    using MerkleProof for bytes32[];
+    uint256 public constant DEFAULT_MAX_INVOCATIONS_PER_ADDRESS = 1;
+    bytes32 public constant CONFIG_MERKLE_ROOT = "merkleRoot";
+    bytes32 public constant CONFIG_USE_MAX_INVOCATIONS_PER_ADDRESS_OVERRIDE =
+        "useMaxMintsPerAddrOverride"; // shortened to fit in 32 bytes
+    bytes32 public constant CONFIG_MAX_INVOCATIONS_OVERRIDE =
+        "maxMintsPerAddrOverride"; // shortened to match format of previous key
+
+    struct ProjectConfig {
+        bool maxHasBeenInvoked;
+        bool priceIsConfigured;
+        // initial value is false, so by default, projects limit allowlisted
+        // addresses to a mint qty of `DEFAULT_MAX_INVOCATIONS_PER_ADDRESS`
+        bool useMaxInvocationsPerAddressOverride;
+        // a value of 0 means no limit
+        // (only used if `useMaxInvocationsPerAddressOverride` is true)
+        uint24 maxInvocationsPerAddressOverride;
+        uint24 maxInvocations;
+        uint256 pricePerTokenInWei;
+    }
+
+    function updateMerkleRoot(
+        mapping(address => mapping(uint256 => bytes32))
+            storage projectMerkleRoot,
+        uint256 _projectId,
+        address _coreContract,
+        bytes32 _root
+    ) internal {
+        require(_root != bytes32(0), "Root must be provided");
+        projectMerkleRoot[_coreContract][_projectId] = _root;
```

Comment:
I think we may want to start exposing an event here, and include it in our interface?

Reply:
the implementation of this emits an event after calling this lib func, https://github.com/ArtBlocks/artblocks-contracts/pull/637/files/88781f12f6fb1a27648c7a4da918feccf6bbacca#diff-28de2faccfae590a400e7771f763d7290f6dc5d5f5fdefd44f6c232a05cba61dR606

Reply:
Oof, I really want the library to emit the event, but it appears that the only way we can accomplish that is for the library to define its own (duplicate) event _in addition_ to the contract defining the same event ([ref](https://medium.com/aragondec/library-driven-development-in-solidity-2bebcaf88736)). I really don't want to go down the path of defining the same event in multiple places, but I also really want these libraries to handle emitting events.

There is also a possibility that we don't use the library to update state here, but instead only use it to `validateMerkleRoot` (in this case, check that it is non-zero), then update state and emit the event in the core contract. In that architecture, I think it would make more sense for the library to not emit the event.

I think this is worth a quick sync.

Reply:
after discussion: deciding to carry-on with current method

lets make it very clear in any library methods where you would expect an event to be emitted to mention that it's on the client (natspec comments)